archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,872 @@
+
+
+## Proof outline
+
+The proof follows: <https://proofwiki.org/wiki/Stirling%27s_Formula>.
+
+We proceed in two parts.
+### Part 1
+We consider the fraction sequence $a_n$ of fractions $n!$ over $\sqrt{2n}(\frac{n}{e})^n$ and
+proves that this sequence converges against a real, positve number $a$. For this the two main
+ingredients are
+ - taking the logarithm of the sequence and
+ - use the series expansion of $\log(1 + x)$.
+
+### Part 2
+We use the fact that the series defined in part 1 converges againt a real number $a$ and prove that
+$a = \sqrt{\pi}$. Here the main ingredient is the convergence of the Wallis product.
+-/
+
+
+open_locale big_operators -- notation ∑ for finite sums
+open_locale classical real topological_space nnreal ennreal filter big_operators
+open  finset filter nat real
+
+namespace stirling
+
+/-- TODO: Perhaps something to add as rat.cast_sum in more generality (see below) in mathlib?!-/
+lemma rat_cast_sum (s : finset ℕ) (f : ℕ → ℚ) :
+  ↑(∑ n in s, f n : ℚ) = (∑ n in s, (f n : ℝ)) :=
+  (rat.cast_hom ℝ).map_sum f s
+-- @[simp, norm_cast] lemma rat_cast_sum [add_comm_monoid β] [has_one β]
+-- (s : finset α) (f : α → ℚ) :
+-- ↑(∑ x in s, f x : ℚ) = (∑ x in s, (f x : β)) := by sorry
+
+/-- **Sum of the Reciprocals of the Triangular Numbers**
+ (copied from archive)
+ TODO: include in some form mathlib or figure out how to import it from archive/ --/
+lemma inverse_triangle_sum :
+  ∀ n, ∑ k in range n, (2 : ℚ) / (k * (k + 1)) = if n = 0 then 0 else 2 - (2 : ℚ) / n :=
+begin
+  refine sum_range_induction _ _ (if_pos rfl) _,
+  rintro (_|n), { rw [if_neg, if_pos]; norm_num },
+  simp_rw [if_neg (succ_ne_zero _), succ_eq_add_one],
+  have A : (n + 1 + 1 : ℚ) ≠ 0, by { norm_cast, norm_num },
+  push_cast,
+  field_simp [cast_add_one_ne_zero],
+  ring,
+end
+
+/-- **Sum of products of consecutive reciprocals** -/
+lemma partial_sum_consecutive_reciprocals :
+ ∀ n, ∑ k in range n, (1 : ℚ) / (k.succ * (k.succ.succ)) ≤ 1 :=
+begin
+  intro n,
+  rw [← (mul_le_mul_left (zero_lt_two)), mul_sum], swap, { exact rat.nontrivial },
+  { have h : ∀ (k : ℕ), k ∈ (range n) →
+      2 * ((1 : ℚ) / (k.succ * (k.succ.succ))) = 2 / (k.succ * (k.succ.succ)),
+      by { intros k hk, rw [mul_div], rw [mul_one (2 : ℚ)] },
+    rw sum_congr rfl h,
+    have h₁ := inverse_triangle_sum n.succ,
+    rw sum_range_succ' at h₁,
+    norm_num at *,
+    rw h₁,
+    rw [sub_le_self_iff],
+    refine (le_div_iff _).mpr (_),
+    { exact (cast_lt.mpr n.succ_pos), },
+    { rw [zero_mul], exact zero_le_two, } },
+ end
+
+
+/-!
+ ### Part 1
+ https://proofwiki.org/wiki/Stirling%27s_Formula#Part_1
+-/
+
+/--
+A sequence of real numbers `an n` has limit `a`, if and only if only if the shifted
+sequence given by `an n.succ` has the limit `a`.
+-/
+lemma tendsto_succ (an : ℕ → ℝ) (a : ℝ) : tendsto an at_top (𝓝 a) ↔
+  tendsto (λ n : ℕ, (an n.succ)) at_top (𝓝 a) :=
+begin
+  nth_rewrite_rhs 0 ← tendsto_map'_iff,
+  have h : map succ at_top = at_top :=
```

Comment:
This seems worthy of a separate lemma, maybe a simp-lemma
```
@[simp] lemma nat.map_succ_at_top : map (succ : ℕ → ℕ) at_top = at_top :=
```
It might even already exist, you could ask on "Is there code for X"?

Reply:
https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/map.20.28succ.20.3A.20.E2.84.95.20.E2.86.92.20.E2.84.95.29.20at_top.20.3D.20at_top